service/task_manager_module.cc

```diff
@@ -0,0 +1,160 @@
+                .scope = info.repair_task_info.is_user_request() ? "table" : "tablet",
+                .state = tasks::task_manager::task_state::running,
+                .keyspace = schema->ks_name(),
+                .table = schema->cf_name()
+            };
+            if (info.repair_task_info.is_user_request()) {
+                // User requested repair may encompass more that one tablet.
+                user_requests[task_id] = std::move(stats);
+                sched_num_sum[task_id] += info.repair_task_info.sched_nr;
+            } else {
+                res.push_back(std::move(stats));
+            }
+            return make_ready_future();
+        });
+
+        for (auto& [id, task_stats] : user_requests) {
+            task_stats.state = sched_num_sum[id] == 0 ? tasks::task_manager::task_state::created : tasks::task_manager::task_state::running;
+            res.push_back(std::move(task_stats));
+        }
+    }
+    // FIXME: Show finished tasks.
+    co_return res;
+}
+
+std::vector<table_id> tablet_virtual_task::get_table_ids() const {
+    return boost::copy_range<std::vector<table_id>>(_ss.get_token_metadata().tablets().all_tables() | boost::adaptors::transformed([] (const auto& table_to_tablets) { return table_to_tablets.first; }));
+}
+
+future<std::optional<tasks::task_status>> tablet_virtual_task::get_status_helper(tasks::task_id id, utils::chunked_vector<locator::tablet_id>& tablets, tasks::virtual_task_hint hint) {
+    auto table = get_table_id(hint);
+    auto schema = _ss._db.local().get_tables_metadata().get_table(table).schema();
+    tasks::task_status res{
+        .task_id = id,
+        .kind = tasks::task_kind::cluster,
+        .is_abortable = co_await is_abortable(std::move(hint)),
+        .keyspace = schema->ks_name(),
+        .table = schema->cf_name(),
+    };
+    size_t sched_nr = 0;
+    auto& tmap = _ss.get_token_metadata().tablets().get_tablet_map(table);
+    co_await tmap.for_each_tablet([&] (locator::tablet_id tid, const locator::tablet_info& info) {
+        if (info.repair_task_info.tablet_task_id.uuid() == id.uuid()) {
+            sched_nr += info.repair_task_info.sched_nr;
+            res.type = locator::tablet_task_type_to_string(info.repair_task_info.request_type);
+            res.scope = info.repair_task_info.is_user_request() ? "table" : "tablet";
+            res.start_time = info.repair_task_info.request_time;
+            tablets.push_back(tid);
+        }
+        return make_ready_future();
+    });
+
+    if (!tablets.empty()) {
+        res.state = sched_nr == 0 ? tasks::task_manager::task_state::created : tasks::task_manager::task_state::running;
+        co_return res;
+    }
+    // FIXME: Show finished tasks.
+    co_return std::nullopt;
+}
+
+table_id tablet_virtual_task::get_table_id(const tasks::virtual_task_hint& hint) const {
+    if (!hint.table_id.has_value()) {
+        on_internal_error(tasks::tmlogger, "tablet_virtual_task hint does not contain table_id");
```

Comment:
If this `table_id` is expected to be always present, then why make it optional?

Reply:
It's always present for tablet virtual tasks, but virtual_task_hint is meant for all virtual tasks

Reply:
Ok.